lab_2_tokenize_by_bpe/main.py

```diff
@@ -58,6 +105,30 @@ def train(
     :param num_merges: required number of new tokens
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(num_merges, int):
+        return None
+    dict_of_token_pairs = count_tokens_pairs(word_frequencies)
+    if dict_of_token_pairs is None:
+        return None
+    list_of_max_token_pairs = []
+    for i in range(0, num_merges):
+        if len(list_of_max_token_pairs) == num_merges:
+            break
+        temp_list_of_max_token_pairs = []
+        max_token_pair = max(dict_of_token_pairs.values())
+        for token_pair in dict_of_token_pairs:
+            if dict_of_token_pairs[token_pair] == max_token_pair:
+                temp_list_of_max_token_pairs.append(token_pair)
+        temp_list_of_max_token_pairs.sort(key=lambda x: (-len(x), x))
+        for temp_token_pair in temp_list_of_max_token_pairs:
+            if len(list_of_max_token_pairs) < num_merges:
```

Comment:
I need an explanation how it can be bigger or equal considering all the checks above

Reply:
I use a cycle (for temp_token_pair in temp_list_of_max_token_pairs:) in the cycle, and I need to check this value every time in this internal cycle, as it can add several values, but cannot have the list length exceed (num_merges) at some point.
That is why I check this value not only in the main but also in the internal cycle

Reply:
Am I correct in my understanding that your approach is to firstly find all pairs to merge and only then perform merging of all of them at once? 
I am afraid it is not exactly correct, as the n+1-th merge may (and probably will) happen for token generated as a result of n-th merge. 
But I appreciate the effort, I get that it's tricky.